File: js/test/acp.test.ts
/**
 * Tests for the ACP (Agent Client Protocol) model source — issue #96, ADR 0025,
 * openspec/changes/add-acp-model-source. Hermetic: no network, no real ACP
 * agent. Drives `test/fixtures/acp-fake-server.mjs`, a small built-ins-only
 * Node script that speaks the same JSON-RPC-over-stdio protocol a real agent
 * (devin acp, Gemini CLI, Zed) does, scripted per ADR 0025's gate items.
 *
 * Run: npm run build && node --experimental-strip-types --test test/acp.test.ts
 */
import { strict as assert } from "node:assert"
import test from "node:test"
import fs from "node:fs"
import os from "node:os"
import path from "node:path"
import { loadACP, type InProcessRequest } from "../dist/index.js"

const FIXTURE = new URL("./fixtures/acp-fake-server.mjs", import.meta.url).pathname

function recordFile(): string {
  return path.join(fs.mkdtempSync(path.join(os.tmpdir(), "acp-test-")), "record.ndjson")
}

function req(text: string, history: string[] = []): InProcessRequest {
  const messages = [...history.map((h) => ({ role: "user", content: h })), { role: "user", content: text }]
  return { messages, model: "acp", body: {} }
}

test("acp: warm session reuse — one process, one session/new, many generate() calls", async () => {
  const record = recordFile()
  const client = await loadACP({ command: "node", args: [FIXTURE], env: { ACP_SCENARIO: "default", ACP_RECORD_FILE: record } })
  try {
    const r1 = await client.generate(req("hello"))
    const r2 = await client.generate(req("again"))
    const r3 = await client.generate(req("third"))
    assert.equal(r1.content, "echo:1:user: hello\nSUPERSEDES-ALL-PRIOR: hello")
    assert.equal(r2.content, "echo:2:user: again\nSUPERSEDES-ALL-PRIOR: again")
    assert.equal(r3.content, "echo:3:user: third\nSUPERSEDES-ALL-PRIOR: third")

    const lines = fs.readFileSync(record, "utf8").trim().split("\n")
    assert.equal(lines.length, 1, "session/new must be sent exactly once across three generate() calls")
  } finally {
    await client.close()
  }
})

test("acp: session/new carries an absolute cwd and an mcpServers array", async () => {
  const record = recordFile()
  const client = await loadACP({ command: "node", args: [FIXTURE], env: { ACP_SCENARIO: "default", ACP_RECORD_FILE: record } })
  try {
    const params = JSON.parse(fs.readFileSync(record, "utf8").trim().split("\n")[0])
    assert.ok(path.isAbsolute(params.cwd), `cwd must be absolute, got ${JSON.stringify(params.cwd)}`)
    assert.ok(Array.isArray(params.mcpServers), "mcpServers must be present as an array")
  } finally {
    await client.close()
  }
})

test("acp: only agent_message_chunk text is accumulated — thought/tool narration is dropped", async () => {
  const client = await loadACP({ command: "node", args: [FIXTURE], env: { ACP_SCENARIO: "noisy" } })
  try {
    const r = await client.generate(req("42"))
    const parsed = JSON.parse(r.content ?? "")
    assert.equal(parsed.answer, "user: 42\nSUPERSEDES-ALL-PRIOR: 42")
  } finally {
    await client.close()
  }
})

test("acp: a permission request is answered, not awaited — the turn completes quickly", async () => {
  const client = await loadACP({ command: "node", args: [FIXTURE], env: { ACP_SCENARIO: "permission" } })
  try {
    const start = Date.now()
    const r = await client.generate(req("delete the database"))
    const elapsed = Date.now() - start
    assert.ok(elapsed < 500, `expected the turn to complete quickly, took ${elapsed}ms`)
    // The fixture offers reject_once first, allow_once second — the client
    // must pick the first option whose kind STARTS WITH "allow".
    assert.equal(r.content, "PERMITTED:allow-once")
  } finally {
    await client.close()
  }
})

test("acp: an unresponsive agent is bounded by the safety-net timeout, not left to hang forever", async () => {
  const client = await loadACP({
    command: "node",
    args: [FIXTURE],
    env: { ACP_SCENARIO: "hang" },
    permissionTimeoutMs: 300,
  })
  try {
    await assert.rejects(client.generate(req("anything")), /timed out/)
  } finally {
    await client.close()
  }
})

test("acp: the supersedes marker prevents a stale answer from a stateful session", async () => {
  const client = await loadACP({ command: "node", args: [FIXTURE], env: { ACP_SCENARIO: "stale" } })
  try {
    const r1 = await client.generate(req("What is the capital of France?"))
    assert.equal(r1.content, "FRESH-ANSWER-TO:What is the capital of France?")

    // Turn 2's full assembled request contains turn 1's text verbatim (the
    // library always sends the WHOLE transcript) — without the supersedes
    // marker a naive stateful agent would match turn 1 first and answer
    // stale. The marker the library appends must make it answer turn 2.
    const r2 = await client.generate(req("What is the capital of Japan?", ["What is the capital of France?"]))
    assert.equal(r2.content, "FRESH-ANSWER-TO:What is the capital of Japan?")
    assert.notEqual(r2.content, "STALE-ANSWER-TO:What is the capital of France?")
  } finally {
    await client.close()
  }
})

test("acp: concurrent generate() calls on one session are serialised, never interleaved", async () => {
  const client = await loadACP({ command: "node", args: [FIXTURE], env: { ACP_SCENARIO: "serialize" } })
  try {
    const p1 = client.generate(req("first"))
    const p2 = client.generate(req("second"))
    const [r1, r2] = await Promise.all([p1, p2])
    assert.equal(r1.content, "done:user: first\nSUPERSEDES-ALL-PRIOR: first")
    assert.equal(r2.content, "done:user: second\nSUPERSEDES-ALL-PRIOR: second")
  } finally {
    await client.close()
  }
})

test("acp: close() is idempotent and actually terminates the child process", async () => {
  const client = await loadACP({ command: "node", args: [FIXTURE], env: { ACP_SCENARIO: "default" } })
  const pid = (client as unknown as { _pid?: number })._pid
  assert.ok(typeof pid === "number" && pid > 0, "expected an internal pid for the test to verify against")

  await client.close()
  await client.close() // must not throw / reject a second time

  assert.throws(() => process.kill(pid as number, 0), /ESRCH/, "the child process should no longer exist")
})

File: elixir/test/acp_test.exs
defmodule Toolnexus.AcpTest do
  @moduledoc """
  ACP (Agent Client Protocol) model source — issue #96, ADR 0031,
  `openspec/changes/add-acp-model-source`. Hermetic: drives a fake ACP
  server (`test/support/fake_acp_server.exs`) over REAL OS pipes, ported
  from `spikes/acp/fakeagent/main.go` / `spikes/acp/SPIKE.md`.
  """
  use ExUnit.Case, async: false

  alias Toolnexus.Acp

  @moduletag timeout: 60_000

  @fixture_script Path.expand("support/fake_acp_server.exs", __DIR__)

  defp fixture_cmd(scenario) do
    jason_ebin = Path.join(Mix.Project.build_path(), "lib/jason/ebin")
    ["elixir", "-pa", jason_ebin, @fixture_script, scenario]
  end

  defp connect!(scenario, opts \\ []) do
    {:ok, pid} = Acp.connect(fixture_cmd(scenario), opts)
    pid
  end

  # -- warm session reuse -----------------------------------------------------

  test "a warm session serves many turns from one process, one session/new" do
    pid = connect!("echo")

    {:ok, a} = Acp.prompt(pid, "first")
    {:ok, b} = Acp.prompt(pid, "second")
    {:ok, c} = Acp.prompt(pid, "third")

    assert a =~ "new_count=1 seq=1"
    assert b =~ "new_count=1 seq=2"
    assert c =~ "new_count=1 seq=3"

    Acp.close(pid)
  end

  test "session/new carries an absolute cwd and an mcpServers array (real devin rejects otherwise)" do
    pid = connect!("echo")
    {:ok, answer} = Acp.prompt(pid, "hi")

    [_, params_json] = String.split(answer, "params=", parts: 2)
    params = Jason.decode!(params_json)

    assert Path.type(params["cwd"]) == :absolute
    assert is_list(params["mcpServers"])

    Acp.close(pid)
  end

  test "generate/1 builds the exact seam create_in_process accepts and drives the in-process client end to end" do
    pid = connect!("echo")
    generate = Acp.generate(pid)

    {:ok, tk} = Toolnexus.create_toolkit(builtins: false)
    client = Toolnexus.Client.create_in_process(model: "acp-fake", generate: generate)
    result = Toolnexus.Client.run(client, "hello", tk)

    assert result.status == "done"
    assert result.text =~ "echo:"

    Acp.close(pid)
  end

  # -- thought/tool narration filtered -----------------------------------------

  test "only agent_message_chunk forms the reply; thought/tool narration is dropped" do
    pid = connect!("noisy")
    {:ok, answer} = Acp.prompt(pid, "hello")

    assert answer == ~s({"answer":"hello"})
    assert {:ok, %{"answer" => "hello"}} = Jason.decode(answer)

    Acp.close(pid)
  end

  # -- permission answered inline, never awaited by the caller -----------------

  test "a session/request_permission is answered with the first allow-kind option, inline, without the caller ever seeing it" do
    pid = connect!("permission")

    {micros, {:ok, answer}} = :timer.tc(fn -> Acp.prompt(pid, "delete the db") end)

    assert answer == "PERMITTED:delete the db"
    # Answered inline from the read loop, not routed through the caller at
    # all — this should complete in well under a second, not hang or wait
    # on any external timeout.
    assert micros < 2_000_000

    Acp.close(pid)
  end

  # -- stale answer prevented by the supersedes marker --------------------------

  test "a stateful session answers a near-duplicate prompt with a STALE answer, unless it carries the supersedes marker" do
    pid = connect!("stale")

    {:ok, a} = Acp.prompt(pid, "What is the capital of France?")
    assert a == "STALE-ANSWER-TO:What is the capital of France?"

    # Full-request-every-turn against a stateful session: the second prompt
    # contains the first, verbatim, with no marker — the fake agent (like a
    # plausible real one) matches the EARLIEST remembered question and
    # answers stale.
    {:ok, b} =
      Acp.prompt(pid, "What is the capital of France?\nWhat is the capital of Japan?")

    assert b == "STALE-ANSWER-TO:What is the capital of France?"

    # The library's own supersedes marker (see Acp.assemble_prompt_text/1)
    # fixes it — the fake agent answers only the text after the marker.
    {:ok, c} =
      Acp.prompt(
        pid,
        "What is the capital of France?\nWhat is the capital of Japan?\n" <>
          "SUPERSEDES-ALL-PRIOR: What is the capital of Japan?"
      )

    assert c == "FRESH-ANSWER-TO:What is the capital of Japan?"

    Acp.close(pid)
  end

  test "generate/1 assembles the supersedes marker itself, so a caller who just appends messages never gets a stale answer" do
    pid = connect!("stale")
    generate = Acp.generate(pid)

    # Even the FIRST turn carries the library-built marker (assembled from
    # every call, not left to the caller), so the fake agent already answers
    # fresh rather than falling back to its naive earliest-match behavior.
    r1 = generate.(%{messages: [%{"role" => "user", "content" => "What is the capital of France?"}]})
    assert r1.content =~ "FRESH-ANSWER-TO:What is the capital of France?"

    r2 =
      generate.(%{
        messages: [
          %{"role" => "user", "content" => "What is the capital of France?"},
          %{"role" => "assistant", "content" => "STALE-ANSWER-TO:What is the capital of France?"},
          %{"role" => "user", "content" => "What is the capital of Japan?"}
        ]
      })

    assert r2.content =~ "FRESH-ANSWER-TO:What is the capital of Japan?"

    Acp.close(pid)
  end

  # -- turn serialisation -------------------------------------------------------

  test "turns on one session are serialised, never sent concurrently" do
    pid = connect!("slow")

    {elapsed_us, results} =
      :timer.tc(fn ->
        [1, 2, 3]
        |> Enum.map(fn n -> Task.async(fn -> Acp.prompt(pid, "t#{n}") end) end)
        |> Enum.map(&Task.await(&1, 10_000))
      end)

    # The fake server sleeps 150ms per prompt; three TRUE-serial turns take
    # >= ~450ms of wall time. If the client let them overlap, this would be
    # closer to 150ms.
    assert elapsed_us >= 400_000

    texts = Enum.map(results, fn {:ok, t} -> t end)
    # Every request got back exactly its own answer — no cross-talk from one
    # session carrying two prompts at once.
    assert Enum.any?(texts, &String.contains?(&1, "t1"))
    assert Enum.any?(texts, &String.contains?(&1, "t2"))
    assert Enum.any?(texts, &String.contains?(&1, "t3"))

    # And the server itself observed them one at a time, in increasing seq
    # order (the client never issued #2 before #1's reply came back).
    seqs =
      texts
      |> Enum.map(fn t ->
        [_, rest] = String.split(t, "slow:", parts: 2)
        [n, _] = String.split(rest, ":", parts: 2)
        String.to_integer(n)
      end)
      |> Enum.sort()

    assert seqs == [1, 2, 3]

    Acp.close(pid)
  end

  # -- process lifetime + idempotent close --------------------------------------

  test "close/1 is idempotent" do
    pid = connect!("echo")
    {:ok, _} = Acp.prompt(pid, "one")

    assert :ok = Acp.close(pid)
    assert :ok = Acp.close(pid)
    assert :ok = Acp.close(pid)
  end

  test "a turn's own failure does not take down the process; the session stays usable" do
    pid = connect!("echo")

    {:ok, _} = Acp.prompt(pid, "first")
    assert Process.alive?(pid)
    {:ok, second} = Acp.prompt(pid, "second")
    assert second =~ "seq=2"

    Acp.close(pid)
  end
end
